lib/vm-watcher.js

```diff
@@ -0,0 +1,341 @@
+    // update the properties of our object now that we've ensured last_modified
+    // is monotonic.
+    props = Object.keys(updateObj);
+    for (propIdx = 0; propIdx < props.length; propIdx++) {
+        prop = props[propIdx];
+        // diff() returns undefined when the properties *are the same*
+        if (diff(knownVm[prop], updateObj[prop])) {
+            knownVm[prop] = updateObj[prop];
+            updated.push(prop);
+        }
+    }
+
+    return (updated);
+}
+
+VmWatcher.prototype.dispatchCreate = // eslint-disable-line
+function dispatchCreate(vmUuid, updateVmobj, watcher) {
+    var self = this;
+    var updated;
+
+    if (self.knownVms.hasOwnProperty(vmUuid)
+        && self.knownVms[vmUuid].destroyed) {
+        // The VM was previously, destroyed: delete since this is a new one.
+        delete self.knownVms[vmUuid];
+    }
+
+    // We only emit if the VM did *not* previously exist. It might have been
+    // noticed by a different watcher first for example.
+    if (!self.knownVms.hasOwnProperty(vmUuid)) {
+        self.emit('VmCreated', vmUuid, watcher);
+        self.knownVms[vmUuid] = {};
+    }
+
+    updated = updateIfNewer(vmUuid, self.log, self.knownVms[vmUuid],
+        updateVmobj);
+
+    if (updated.length > 0) {
+        self.log.trace({
+            event: 'create',
+            modifiedFields: updated,
+            vm: vmUuid,
+            watcher: watcher
+        }, 'VM has changed');
+    }
+};
+
+VmWatcher.prototype.dispatchModify = // eslint-disable-line
+function dispatchModify(vmUuid, updateVmobj, watcher) {
+    var self = this;
+    var SPECIAL_STATES = ['failed', 'provisioning'];
+    var updated;
+
+    if (!self.knownVms.hasOwnProperty(vmUuid)) {
+        self.knownVms[vmUuid] = {};
+    }
+
+    if (self.knownVms[vmUuid].destroyed) {
+        // We don't modify destroyed VMs. Wait for a create first.
+        self.log.warn({
+            updateVmobj: updateVmobj,
+            vm: vmUuid,
+            watcher: watcher
+        }, 'Skipping modify for already destroyed VM ' + vmUuid);
+
+        return;
+    }
+
+    if (watcher !== 'periodic'
+        && SPECIAL_STATES.indexOf(self.knownVms[vmUuid].state) !== -1) {
+        // when we're in one of the 'special' states, zone_state and state don't
+        // match. Since watchers other than periodic don't go through vmadm,
+        // they won't know about the correct 'state' in these cases, so we leave
+        // it to the periodic watcher to update the state for them.
+        delete updateVmobj.state;
+    }
+
+    updated = updateIfNewer(vmUuid, self.log, self.knownVms[vmUuid],
+        updateVmobj);
+
+    if (updated.length > 0) {
+        self.log.trace({
+            event: 'modify',
+            modifiedFields: updated,
+            vm: vmUuid,
+            watcher: watcher
+        }, 'VM has changed');
+
+        self.emit('VmModified', vmUuid, watcher);
```

Comment:
We should emit events on the next turn of the event loop, using `setImmediate` like following:

```
setImmediate(function() { self.emit('VmModified', vmUuid, watcher); });
```
so that event handlers are not called synchronously. There are several problems with emitting events synchronously:

1. If an event is emitted as part of the setup of the event emitter, the caller that creates the event emitter will not have time to set an event listener before the event is emitted.

2. It makes the program handle events as soon as they are emitted, which can have various undesirable effects. One of them is preventing other async operations from completing until the event has ben completely handled, which has a negative impact on performance.

3. It's not in line with the best practice used mostly by all node core event emitters of emitting events asynchronously.